broadinstitute/single_cell_portal_core
# Fixing scatter plot legends in spatial UX (SCP-5626)

#### BACKGROUND & CHANGES
This update fixes issues discovered with scatter plot legends when viewing multiple plots in the spatial UX.  Whenever a user selects a new spatial plot, every legend shown then updates the counts to use data from the newest spatial plot.  Additionally, if certain labels are missing from the newest plot, those legend entries are still shown but with a count of 0.

The cause was from a shared `useState` reference in `ExploreDisplayTabs` for setting label counts (this was done so it could be shared with various components in the differential expression UX).  Now, counts are managed at the `ScatterPlot` level instead of `ExploreDisplayTabs`.  Additionally, color assignments are now tracked using the main, non-spatial plot as the control.  Any additional spatial plots will use that color map to ensure that labels keep the same color assignment across all plots.

Note: you may see a brief flicker as colors are reassigned, this has to do with waiting for the main plot to finish rendering.

#### MANUAL TESTING
The easiest way to test this PR is to ingest data from [SCP1375](https://singlecell.broadinstitute.org/single_cell/study/SCP1375) (this is the study where the above bugs were found).  Only 4 files are needed for testing:
- [metadata.csv](https://singlecell.broadinstitute.org/single_cell/data/public/SCP1375/integrative-in-situ-mapping-of-single-cell-transcriptional-states-and-tissue-histopathology-in-an-alzheimer-disease-model?filename=metadata.csv) (Metadata file)
- [top_level_umap.csv](https://singlecell.broadinstitute.org/single_cell/data/public/SCP1375/integrative-in-situ-mapping-of-single-cell-transcriptional-states-and-tissue-histopathology-in-an-alzheimer-disease-model?filename=top_level_umap.csv) (Normal clustering)
- [spatial_8months-control-replicate_1.csv](https://singlecell.broadinstitute.org/single_cell/data/public/SCP1375/integrative-in-situ-mapping-of-single-cell-transcriptional-states-and-tissue-histopathology-in-an-alzheimer-disease-model?filename=spatial_8months-control-replicate_1.csv) (Spatial transcriptomics)
- [spatial_13months-disease-replicate_2.csv](https://singlecell.broadinstitute.org/single_cell/data/public/SCP1375/integrative-in-situ-mapping-of-single-cell-transcriptional-states-and-tissue-histopathology-in-an-alzheimer-disease-model?filename=spatial_13months-disease-replicate_2.csv) (Spatial transcriptomics)

1.  Boot all services and sign in
2. Create a new study and ingest the 4 files above, making sure to upload the last two cluster files as spatial
3. Once all clustering/metadata have ingested, go to the Explore tab for this study (click "View study" at the top left of the upload wizard)
4. Load `top_level_umap.csv` and `top_level_cell_type` as the annotation
5. In the Spatial groups menu, select `spatial_8months-control-replicate_1.csv`
6. Confirm the count for `Astro` is **6789** for `top_level_umap.csv` and **919** for `spatial_8months-control-replicate_1.csv`
7. Confirm that `spatial_8months-control-replicate_1.csv` does not have the group `LHb`, and the color for `Micro` and the remaining labels match in both plots
8. Add `spatial_13months-disease-replicate_2.csv` from the spatial groups menu
9. Confirm the count for `Astro` is **927** and that it has the `LHb` group with a count of **65** and the correct color 
10. Change the annotation multiple times (including back to a value you've already selected) and confirm that color assignments match across all plots.
    * If you select `Hide all` in the legend for the main plot then scroll to the bottom and select the last entry, this is a quick way to ensure that assignments have stayed consistent across all plots.
